cdragontoolbox/sknfile.py

```diff
@@ -0,0 +1,71 @@
+from .tools import BinParser
+
+
+class SKNFile:
+    def __init__(self, file):
+        if isinstance(file, str):
+            file = open(file, "rb")
+        if file.read(4) != b"\x33\x22\x11\x00":
+            raise ValueError("missing magic code")
+
+        f = BinParser(file)
+
+        self.major, self.minor, self.count = f.unpack("<HHI")
+        self.entries = [self.read_object(f) for i in range(self.count)]
+
+        if self.major == 4:
+            self.unknown, = f.unpack("<I")
+
+        self.index_count, self.vertex_count = f.unpack("<II")
+
+        if self.major == 4:
+            self.vertex_size, = f.unpack("<I")
+            self.contains_tangent = bool(f.unpack("<I")[0])
+            self.bounding_box_min = f.unpack("<fff")
+            self.bounding_box_max = f.unpack("<fff")
+            self.bounding_sphere_location = f.unpack("<fff")
+            self.bounding_sphere_radius, = f.unpack("<f")
+
+        self.indecies = [f.unpack("<H")[0] for i in range(self.index_count)]
+        self.vertices = [self.read_vertex(f) for i in range(self.vertex_count)]
+
+        for entry in self.entries:
+            entry["vertices"] = self.vertices[entry["start_vertex"] : entry["start_vertex"] + entry["vertex_count"]]
+            entry["indecies"] = self.indecies[entry["start_index"] : entry["start_index"] + entry["index_count"]]
+            entry["indecies"] = [x - (0 if x > entry["start_vertex"] else entry["start_vertex"]) for x in entry["indecies"]]
+
```

Comment:
- `vertices` and `indecies` are temporaries
- You should not set then override `entry["indecies"]`. Just use temporary and only set the final, true value.
- You don't need `start_index`, `start_vertex`, etc. in the final entries. You should parse them in `read_object()` but keep on `self` entries with only the relevant, non-redundant fields; something like this:
```py
entries_info = [self.read_object(f) for i in range(self.count)]
...
self.entries = []
for info in entries_info:
    self.entries.append({"name": info["name"], "vertices": ..., ", "indices": ...})
```